kolena/_experimental/object_detection/utils.py

```diff
@@ -177,3 +177,85 @@ def compute_confusion_matrix_plot(
     for actual_label in ordered_labels:
         matrix.append([confusion_matrix[actual_label][predicted_label] for predicted_label in ordered_labels])
     return ConfusionMatrix(title=plot_title, labels=ordered_labels, matrix=matrix)
+
+
+def _compute_sklearn_arrays_by_class(
+    all_matches: List[MulticlassInferenceMatches],
+) -> Tuple[np.ndarray, np.ndarray, Dict[str, np.ndarray], Dict[str, np.ndarray]]:
+    y_true: List[int] = []
+    y_score: List[float] = []
+    y_true_by_label: defaultdict[str, List[int]] = defaultdict(lambda: [])
+    y_score_by_label: defaultdict[str, List[float]] = defaultdict(lambda: [])
+    for image_bbox_matches in all_matches:
+        for _, bbox_inf in image_bbox_matches.matched:  # TP (if above threshold)
+            y_true.append(1)
+            y_score.append(bbox_inf.score)
+            y_true_by_label[bbox_inf.label].append(1)
+            y_score_by_label[bbox_inf.label].append(bbox_inf.score)
+        for gt_or_pair in image_bbox_matches.unmatched_gt:  # FN
+            gt_label: str = gt_or_pair[0].label if type(gt_or_pair) is tuple else gt_or_pair.label
+            y_true.append(1)
+            y_score.append(-1)
+            y_true_by_label[gt_label].append(1)
+            y_score_by_label[gt_label].append(-1)
+        for bbox_inf in image_bbox_matches.unmatched_inf:  # FP (if above threshold)
+            y_true.append(0)
+            y_score.append(bbox_inf.score)
+            y_true_by_label[bbox_inf.label].append(0)
+            y_score_by_label[bbox_inf.label].append(bbox_inf.score)
+
+    y_true_by_label_np: Dict[str, np.ndarray] = {}
+    for key, value in y_true_by_label.items():
+        y_true_by_label_np[key] = np.array(value)
+
+    y_score_by_label_np: Dict[str, np.ndarray] = {}
+    for key, value in y_score_by_label.items():
+        y_score_by_label_np[key] = np.array(value)
+
+    return np.array(y_true), np.array(y_score), y_true_by_label_np, y_score_by_label_np
+
+
+def _compute_optimal_f1_with_arrays(
+    y_true: np.ndarray,
+    y_score: np.ndarray,
+) -> float:
+    if np.all(y_true == 0):
+        return 0.0
+
+    precision, recall, thresholds = sklearn_metrics.precision_recall_curve(y_true, y_score)
+
+    # delete last pr of (1,0)
+    precision = precision[:-1]
+    recall = recall[:-1]
+
+    if thresholds[0] < 0:
+        precision = precision[1:]
+        recall = recall[1:]
+        thresholds = thresholds[1:]
+    if len(thresholds) == 0:
+        return 0.0
+
+    f1_scores = 2 * precision * recall / (precision + recall)
+    f1_scores = np.nan_to_num(f1_scores, nan=0)
+    max_f1_index = np.argmax(f1_scores)
+    if f1_scores[max_f1_index] == 0:
+        return 0.0
+    else:
+        return float(thresholds[max_f1_index])
+
+
+def compute_optimal_f1(
```

Comment:
There isn't much utility in providing a single function for both single-class and multi-class here, and it forces users of this to deal with an annoying union return type. Would prefer to export separate `compute_optimal_f1_score` and `compute_optimal_f1_score_multiclass` functions